dnsmasq/patches/320-prefered-server-strategy.patch

```diff
@@ -0,0 +1,67 @@
+diff --git a/src/forward.c b/src/forward.c
+index 942b02d..af75209 100644
+--- a/src/forward.c
++++ b/src/forward.c
+@@ -306,6 +306,8 @@ static int forward_query(int udpfd, union mysockaddr *udpaddr,
+ 	  return 1;
+ 	}
+ #endif
++      /* OVH(prefered_dns) : on query timeout we disable strict order mode, that make next query to be sent to all servers */
++      reset_option_bool(OPT_ORDER);
+ 
+       /* retry on existing query, send to all available servers  */
+       domain = forward->sentto->domain;
+@@ -739,6 +741,8 @@ void reply_query(int fd, int family, time_t now)
+   ssize_t n = recvfrom(fd, daemon->packet, daemon->packet_buff_sz, 0, &serveraddr.sa, &addrlen);
+   size_t nn;
+   struct server *server;
++  /* OVH(prefered_dns) : a server pointer to store the prefered server */
++  struct server *prefered=NULL;
+   void *hash;
+ #ifndef HAVE_DNSSEC
+   unsigned int crc;
+@@ -761,13 +765,19 @@ void reply_query(int fd, int family, time_t now)
+   
+   /* spoof check: answer must come from known server, */
+   for (server = daemon->servers; server; server = server->next)
+-    if (!(server->flags & (SERV_LITERAL_ADDRESS | SERV_NO_ADDR)) &&
+-	sockaddr_isequal(&server->addr, &serveraddr))
+-      break;
+-  
++    if (!(server->flags & (SERV_LITERAL_ADDRESS | SERV_NO_ADDR)))
++    {
++      /* OVH(prefered_dns) : the prefered server is the first in the resolv.conf list */
++      if(prefered == NULL && !(server->flags & (SERV_HAS_DOMAIN))) {
```

Comment:
SERV_HAS_DOMAIN flag is checked here because --server and --rev-server are in server's list before the "resolv.conf".